### ISSUE TITLE
RFC: Improve Error Handling

### ISSUE THREAD

**Issue by bfitch** [cerebral/cerebral#731]
One of the main goals of Cerebral is to understand the "flow" of your application. *Signals* give us a nice high level view of our app logic, but often times error handling clutters and adds noise to our otherwise beautiful signals. 

A key design goal for error handling should be, in my opinion, that you only need to explicitly handle errors where and when your program requires it. Otherwise, error handling actually *detracts* from the readability of your program, cluttering it with low level book keeping instead of expressing your intent (which is what Cerebral is all about).

### The Problem

Take this (Cerebral 1) signal for example:
```js
module.addSignals({
  mounted: [
    setupDeviceSettings,
    setAuthHeaders,
    fetchCurrentUser, {
      success: [
        copy('input:currentUser', 'state:currentUser'),
        copy('state:currentUser.current_account_id', 'state:navbar.newPost.account_id'),
        fetchPosts, {
          success: [
            copy('input:posts', 'state:posts'),
            when('flags.showCommentsButton'), {
              true: [
                fetchComments, {
                  success: [copy('input:comments', 'state:comments')],
                  error: [(e) => { console.log('comments error', JSON.stringify(e)); }]
                }
              ],
              false: [/*noop*/]
            }
          ],
          error: [(e) => { console.log('fetch posts error', JSON.stringify(e)); }]
        },
      ],
      error: [(e) => { console.log('fetchCurrentUser error', JSON.stringify(e)); }]
    }
  ]
})
```

This signal is really just about fetching some related resources from some APIs. We need to get the: `currentUser`, their `posts` and finally the `comments` for their posts. We can't render the first view until we have that data. 

Notice how repetitive the error handling is above and the "pyramid of doom" with indentation creeping to the right of the screen. In this case, the developer is forced to define the error path for each API call and provide an explicit error handling action. But, really, all we care about in this instance is: 1) did everything work? or 2) did something prevent us from getting this data?

### The Solution

As we talked about in the hangout, I'd like to start a discussion around a better error handling solution for Cerebral. What if we could re-write the signal above like (please forgive this hodgepodge of Cerebral 1 and 2 syntax 😛) :

```js
module.addSignals({
  mounted: sequence([
    setupDeviceSettings,
    setAuthHeaders,
    fetchCurrentUser, {
      success: [
        copy('input:currentUser', 'state:currentUser'),
        copy('state:currentUser.current_account_id', 'state:navbar.newPost.account_id'),
        fetchPosts, {
          success: [
            copy('input:posts', 'state:posts'),
            when('flags.showCommentsButton'), {
              true: [
                fetchComments, {
                  success: [copy('input:comments', 'state:comments')]
                }
              ],
              false: [/*noop*/]
            }
          ]
        }
      ]
    }
  ], {
    error: [flash('Error fetching data')]
  })
})
```

The idea is to provide a named path at the top level that can be invoked if any of the nested paths throw an error. Conceptually, a nested error will "bubble up" to the top of the signal chain until it is handled. In the event that you want more specific error handling, you simply define error paths like in the first example. The key thing is that now you are not *forced* to repeat yourself with error handling when your application doesn't require you to.  Already, with just this one change, I think the logical flow is clearer for "the next developer". You can more easily scan the signal without mentally removing the error paths. 

It's always a difficult balance between being low level + explicit vs concise/abstract. In this case, however, I think handling errors more concisely would benefit from a higher level API and convention.

Anyway, that's all I have for now. Everyone please weigh in with ideas, excitement, or objections. 😄

**Comment by christianalfoni** [comment]
I definitely think this makes sense. The exact details of the implementation is a bit trickier though.

1. When you say "bubble up", should it rather "go to" the "global" paths? Or should it check parent paths for "same named" paths?

2. How would we visualize this in the debugger?

3. What happens when you compose sequences?

4. Can a "global" path trigger an other global path?

Here is what I suggest:

1. When you define a sequence, or parallel you can define default paths for them. That means the `path` object in actions will always have these paths attached to them. They are attached to the sequence, so only actions defined within the sequence can use those paths

2. It shows these extra paths on all paths of the sequence. So every action defining some paths after them will have these extra path names attached. If they are hit, it will show it is the executed path. Basically it just hides the fact that these are defined separate. It makes everything easier to implement as well, as the default paths can just be `Object.assigned` to any paths found in the sequence

3. Default paths or only for the specific sequence/parallel, or you will have to jump files to understand how these work, which would cause havoc I think

4. Difficult one... I think "no". This also fits with the techical implementation of just `Object.assign` default paths to paths in the sequence

**Comment by gaspard** [comment]
As an extra step, I would consider adding a 'continue' thing that would run the next action:

```js
export const  mounted = sequence([
  setupDeviceSettings,
  setAuthHeaders,
  fetchCurrentUser, 
  set(props`currentUser`, state`currentUser`),
  set(state`currentUser.current_account_id`, state`navbar.newPost.account_id`),
  fetchPosts,
  set(props`posts`, state`posts`),
  when(state`flags.showCommentsButton`), {
    true: [
      fetchComments,
      set(props`comments`, state`comments`)
    ],
    false: [/*noop*/]
  }
], {
  success: sequence.continue,
  error: [flash('Error fetching data')]
})
```

**Comment by bfitch** [comment]
Great! Glad you think it makes sense @christianalfoni :) 

1. I was originally thinking the default error path would be on a per signal basis, but I think per `sequence` (or `parallel`) is just as good.

2. I was originally thinking the error path would appear at the top level of the sequence, like where it's defined in the signal code. But, if attaching to each action's path is easier to implement, I say we roll with that and then make a change only if there's some usability problem. Better to iterate rather the guess which design is best.

3. I think having default paths only active per sequence makes a lot of sense. It would be "nice to have" if a parent sequence's default `error` could be called if a chid composed `sequence` throws (without its own error handling). BUT, having each sequence responsible for setting a default is good too, I think.

4. I agree: no. Default paths should only be per sequence, no cross signal "go to". :)

Regarding "bubbling up", I think of it the same as DOM event handlers. You can attach a handler to a nested <button> or a parent <div> or all the way up at `window`. The nested handler, if defined will get called first, but if not defined, the event will continue up until the `window` scope. I imagine the same semantics for cerebral errors. If an error occurs in a nested path, any locally defined error paths will get called and handle the error. But, if none are found, the error will continue to bubble until it reaches the top of the `sequence` where the default path can handle it.

Hope that makes sense and thanks for weighing in. 😛

**Comment by bfitch** [comment]
@gaspard I had the same thought right after I posted! It seems like if we handle the default error case, why not handle the default success case as well? It cleans up the signal so much.

**Comment by henri-hulski** [comment]
@bfitch Sounds similar to exception handling at least in Python. The exception 'bubbles up' until it gets caught / excepted. 
Then you can handle it.

**Comment by christianalfoni** [comment]
I suggest I do a simple `Object.assign` implementation of this first, so:

```js
sequence([
  actionA, {
    success: [actionB]
  }
], {
  error: []
})
```

I agree with @gaspard that this does not give much value until you can remove that paths alltogether.

```js
sequence([
  actionA,
  actionB
], {
  error: []
})
```

The problem is that this is invalid. There is no paths to go down on `actionA`, so the `path` argument will not even be there inside `actionA`.  Also the fact that you probably do not want to move further if `actionA` fails. It does not describe that.

So I think this last example is more of an "abort" pattern, rather then "default paths". Hm hm... this needs a lot of thinking :)

**Comment by bfitch** [comment]
Let's pretend that we don't know anything about cerebral or paths...if I'm just looking at the sequence, this "abort pattern" is exactly what I would expect to happen.

Run actionA, then run actionB- only if actionA succeeds. If there are errors call the error handler.

So with this approach, you only explicitly supply paths if your flow truly requires something more complicated than "did this succeed or did this fail?"

So I agree this might be terrible from an implementation perspective haha! But, from a users perspective...I feel like it is exactly right and will improve the readability and usability of signals greatly. Very exciting!

**Comment by bfitch** [comment]
Another nice thing is the sequence actually looks like a sequence of steps, rather than a tree branching out in multiple directions 😀

**Comment by pbit** [comment]
> Let's pretend that we don't know anything about cerebral or paths...

...`then` I would expect to `catch` errors.

This would align error handling with promises.
Catch at low level or let it bubble up.

**Comment by christianalfoni** [comment]
Also to remember here is that function-tree does not throw an error on `error` path... it is just a path. Paths are not related to errors at all actually. Errors are caught inside actions and you choose some random path name to go down. `error` just happens to be a typical path... it is not an error in itself :)

Another thing to consider is that function-tree is static, meaning that we can not make it run differently based on some scenario. We have to analyze it before hand (to produce possible paths to go down), verify that the tree is correct and also be able to describe visually how it will execute before it actually does.

But we are discussing two different solutions:

## Default paths

```js
sequence([
  someAction, {
    baz: []
  }
], {
  foo: [],
  bar: []
})
```

In this solution we give default paths, meaning that the `{foo: [], bar: []}` object would be merged into any object inside the sequence, meaning that `{foo: [], bar: [], baz: []}` would become the paths. This keeps the tree static.

This solution does not "flatten" the execution in any way though, but if you use a lot of error paths, it can clean up the tree a bit. Personally I do not think this is really what we are after :)

## Abort execution
Currently you can abort execution using `context.execution.abort()`. This is a low level implementation that just throws an error and an `abort` event, making the tree execution just stop. The tree itself is still static, it just stops running.

What we are talking about here is making the tree run dynamically, jumping arbitrary to some paths based on any execution errors. This is very hard to do with current implementation.

To make this work we would have to trigger a new signal. Like:

```js
function someAction ({abort}) {
  return abort.error()
}

sequence([
  someAction
], {
  error: []
})
```

Technically this is what would happen:

```js
controllerFunctionTree.on('abort', (execution) => {
  treeToRun = execution.addTheErrorPathSomehow

  controller.runSignal('abort.error', treeToRun)
})
```

Meaning that the debugger would then show a brand new signal when this occurs.

I have a feeling this is not what we want either :)

### Inject paths
Like, @gaspard example, is definitely in the area we really want to go with this. So a third option would be to inject paths. That means:

```js
sequence([
  setupDeviceSettings,
  setAuthHeaders,
  fetchCurrentUser, 
  set(props`currentUser`, state`currentUser`),
  set(state`currentUser.current_account_id`, state`navbar.newPost.account_id`),
  fetchPosts,
  set(props`posts`, state`posts`),
  when(state`flags.showCommentsButton`), {
    true: [
      fetchComments,
      set(props`comments`, state`comments`)
    ],
    false: [/*noop*/]
  }
], {
  error: [flash('Error fetching data')]
})
```

Function-tree would inject `{error: []}` behind every function that does not already have a set of paths. Meaning it would look like this when evaluated:

 ```js
sequence([
  setupDeviceSettings, {error: []},
  setAuthHeaders, {error: []},
  fetchCurrentUser, {error: []}, 
  set(props`currentUser`, state`currentUser`), {error: []},
  set(state`currentUser.current_account_id`, state`navbar.newPost.account_id`), {error: []},
  fetchPosts, {error: []},
  set(props`posts`, state`posts`), {error: []},
  when(state`flags.showCommentsButton`), {
    true: [
      fetchComments, {error: []},
      set(props`comments`, state`comments`) {error: []},
    ],
    false: [/*noop*/]
  }
], {
  error: [flash('Error fetching data')]
})
```

Then we have to allow function-tree to skip paths. Meaning the when for example `setupDeviceSettings` returns no path, it means that it should continue to `setAuthHeaders`. This is perfectly okay in the "static world", cause the tree does not change, only the way it traverses it.

The "bad thing" is that in the debugger it would show this error path on every action. The good thing though is that if an error occurs it will actually show it explicitly, what action actually caused it.

Another benefit is that you can choose if the error should cause the tree to stop running.

 ```js
sequence([
  setupDeviceSettings, {error: []},
  setAuthHeaders, {error: []},
  fetchCurrentUser, {error: []}, 
  set(props`currentUser`, state`currentUser`), {error: []},
  set(state`currentUser.current_account_id`, state`navbar.newPost.account_id`), {error: []},
  fetchPosts, {error: []},
  set(props`posts`, state`posts`), {error: []},
  when(state`flags.showCommentsButton`), {
    true: [
      fetchComments, {error: []},
      set(props`comments`, state`comments`) {error: []},
    ],
    false: [/*noop*/]
  }
], {
  error: [function ({execution}) {execution.abort()} ]
})
```

Maybe you do not want it to stop running, but rather list up things the failed etc.

**Comment by gaspard** [comment]
@christianalfoni I agree.

We should consider this as **syntactic sugar** to ease writing and maintaining code, not as some new feature to create jumps and hoops.

To make things a tad nicer, we could have an "abort" operator that could take a message to be displayed in the debugger at the end of the chain (something similar to what you show on a thrown error or maybe even the same thing).

:-)

**Comment by christianalfoni** [comment]
@gaspard So you thinking that we should keep as is, but rather make `abort` a thing? So when an `abort` happens the debugger is informed? We can do the same as with error. When error happens now the action in question is highlighted. We can do same with abort. Highlight the action that caused the abort :)

**Comment by gaspard** [comment]
There are two things to this "pyramid of doom" thing:

1. Ease writing cleaner code without nesting
2. Ease feedback on abort

For (2), yes @christianalfoni, the idea of an "abort" solves this. We can do this now by throwing an error but I think "abort" should not be a red box thing. It's not an error, it's normal code execution.

For (1), I really like the idea that the "sequence" function prepares the tree by merging or adding the default paths. The only issue I see is for actions not using "path". For this situation, I propose this:

```js
// sequence
[
  actionA, { foo: [], bar: [] },
  actionB
]
```

If **actionA** does not return a path, it jumps directly to actionB. This simple change also allows writing actions that simply continue on success and move into the brace on error.

**Comment by tiagocpontesp** [comment]
This seems to me more of a ui-handling type of thing. 

I think it would be cool that in cerebral-land, a supported signal's "path" could be hooked up to a module/controller, e.g. could have an error widget module, with chains hooked up to their respective error types.

Better yet if the system made use of caught Errors from the execution promises. That is the main function-tree error handling issue imho.

Aside from uncaught errors that you probably want to handle in this same module, I see any kind of aborting execution paths as custom app errors. 

I'm suggesting everything based on the assumption that these paths can/should be moved elsewhere conceptually in the app source, in the process leaving the concept of "paths" to the "happy" app logic only.

**Comment by christianalfoni** [comment]
Yeah, we can combine these concepts. For example:

1. We implement `abort` event to work the same way as `error` even. It highlights the action that caused the abort

2. We can create a module that handles different types of aborts. It just adds an `abort` event listener and triggers signals 

```js
modules: {
  aborts: AbortModule([
    // Just show a snackbar
    set(state`snackbar`, props`error.message`)
    // Check type of error
    equals(props`error.type`), {
      TypeError: [],
      otherwise: []
    }
  ])
}
```

Basically it listens to `abort` event where you get the payload of the abort, along with whatever else is on the current payload. So could do:

```js
function someAction ({abort}) {
  return abort({error: new AppError('something failed')})
}

{
modules: {
  aborts: AbortModule([
    equals(props`error.type`), {
      AppError: [],
      otherwise: []
    }
  ])
}
```

**Comment by bfitch** [comment]
I think this all sounds great. But, just to clarify...we're talking about two things. There's (1) the API as outlined by @gaspard and (2) the implementation, mostly proposed by @christianalfoni. 

I think what we all want, API and behavior wise, is what @gaspard outlined above. But, just to make sure, aside from implementation details, this is my summary:

- A sequence is a list of actions
- Actions are executed in order
- Actions, by default, execute along a "happy path", passing output from one action to the next until the end of the sequence
- *If* there is a need to stop the execution and "bail out" of the sequence, because of either: (1) An error or (2)`abort` is explicitly called, execution can "go to" a defined `path` at the action level all the way up to the top level of the sequence. (this maybe is not totally decided...)
- Actions can also define custom `paths` if branching logic is required in the sequence

That's my understanding of The Concept™. Do you all agree with that? Did I miss something? 

Let's also continue the implementation discussion. But wanted to make sure we're on the same page with what we want. :)

**Comment by fopsdev** [comment]
@bfitch pretty much how i understood it as well.
beeing used to try-catch i could imagine something like:
```js
sequence('A Sequence', [
ActionA,
ActionB
],{
error: [
HandleError
]
abort: [
HandleAbort
]
}
```
might look familliar to us because the last {} in the squence-method looks like a catch block?

Love  The Concept™ btw ;)

**Comment by bfitch** [comment]
Yes @fopsdev that looks very close to what I was imagining. The only nuance would be that you could define an error path at the action level as well:

```js
sequence('A Sequence', [
  ActionA, { error: [handleErrorInActionA] },
  ActionB,
  ActionC
], {
error: [HandleError] // handle error in ActionB, ActionC (or ActionA if no nested error path defined
abort: [HandleAbort]
})
```

**Comment by gaspard** [comment]
@bfitch I would also like to add that these "error", "abort" paths have nothing special about them. 

There could also be things like this:

```js
sequence('A Sequence', [
  ActionA, { error: [handleErrorInActionA] },
  ActionB,
  ActionC
], {
toast: [showAToast] // this chain is sometimes executed by "ActionB"
error: [handleError, abort]
})
```

This above is "compiled" as:

```js
sequence('A Sequence', [
  ActionA, { error: [handleErrorInActionA], toast: [showAToast] },
  ActionB, { error: [handleError, abort] }
  ActionC, { error: [handleError, abort], toast: [showAToast] }
])
```

All the default paths are added or merged after each action. If ActionC does not return a path, it just ignores the braces (see previous comment)...

**Comment by christianalfoni** [comment]
@gaspard I think one important ingredient you add to this example is:

```js
{
toast: [showAToast],
error: [handleError, abort] // <- ABORT
}
```

The really good thing here is that we do not just abort, we actually compose in "How to handle errors". Maybe sometimes we just want some default behaviour but let the signal continue or we actually want to abort current execution. That is a 👍 

What is a bit risky here is that we do allow a very implicit approach to paths. There is nothing in the code that indicates this "compiling", and you would also in the debugger risk having quite a few paths that is never used by actions. It might actually get super messy.

So what I think we need to question is: "Do we really need default paths? Or is this really about handling that something went wrong in a generic way?". And by "something went wrong" I mean that there is no reason to continue signal execution.

I think what we really want is the second here. We want to "jump out" of an executing signal and alternatively do something related to that because there is no need to continue execution. Having default paths is an interesting idea, but I am afraid it is a "lazy" thing. It saves some code, but it reduces the explicitness of the signal as the paths will be hidden.

So here is another suggestion:

```js
sequence([
  getThis, // Attaches some payload to props
  getThat, // Attaches some other payload to props
  thenGetThis, // And another
  computeItAll // Here we handle it all
], [
  showToast('baaad stuff')
])
```

With this solution we just add an additional indication that "when this signal aborts for any reason, run this sequence". It allows us to have a pattern where the actions have no paths they go down, you just extract stuff and put it on props until you are ready to handle it. It is basically the same as a bunch of `thens` and a `catch` at the end. It is also similar to observables that has two initial callbacks. One for success and one for error. We are about chains, so we have one for everything okay and one for bailout. 

Now... this causes an implementation problem, because there is no such thing as "abort and go down this path", there is neither any existing "visual hook" for this in the debugger. That said, an `abort` does give information about where it happened, meaning that we could indicate an `abort` and what happened specifically in the debugger. Just needs to be implemented.

`then` + `catch` and two observable callbacks are proven concepts. I am afraid default paths will cause us unknown problems and some concerns are already raised. Overloading sequences with paths, making mess in debugger, and making sequences become completely implicit, making it harder to understand how they run by looking at the code.

This was a lot of text, sorry, but I hope it makes sense :)

**Comment by bfitch** [comment]
I agree @christianalfoni. The primary value is a clearer API for executing the "happy path" in the chain while also handling "catch" for an error/abort. Awesome.

**Comment by christianalfoni** [comment]
You know, `abort` was not a very "thought through" implementation. After this discussion it becomes rather apparent that this should work like normal error throwing:
 
```js
// Will cause catch sequence to run
function action () {
  throw 'it failed'
}

// Will cause catch sequence if error thrown since no `catch`
// is defined
function asyncAction ({http}) {
  return http.get()
}

sequence([
  action,
  asyncAction
], [
  // Catch thrown errors
])
```

This makes A LOT of sense I think :)

**Comment by henri-hulski** [comment]
In Python we use the EAFTP (Easier to ask for forgiveness than permission) coding style. 
So e.g. for check if keys or attributes are valid we use try/except blocks instead of if statements. 
You can just throw somewhere

**Comment by gaspard** [comment]
@christianalfoni My main motivation for a generic "default" paths was to avoid giving a special meaning to "success" path. I also thought of using throw/catch but it felt a tad less "interesting" but as you said, it could be much better.

The only things I would add is that we must discriminate between **runtime errors** and **flow cancellation errors**. The first type of errors should never occur and should not be catched (at least at this level) as they are signs of ill-working code and might not be seen during testing. The second type of errors are things like "wrong password", "unknown user" and should have some kind of payload so that the error handling code can make something useful out of it.

So I think we must throw a specific kind of error with a payload:

```js
throw sequence.abort(`blah blah`, { /* some payload that will be stringyfied */ })
```

**Comment by fopsdev** [comment]
yeah and what do you think of doing smth like:
```js
throw new error(path: 'abort', payload: {})
```
and then function-tree could somehow magically ;) locate the closest block which contains a ```{ abort: [] }```

**Comment by christianalfoni** [comment]
@gaspard Hm, yeah... this was a great example to discuss on. So in my mind, if an error is wrong, you would:

```js
[
  login, {
    success: [],
    invalidUser: []
  }
]

// Inside login

function login ({http, path}) {
  return http.post(credentials)
    .then(path.success)
    .catch(response => {
      if (response.status === 400) { return path.invalidUser() }

      throw response.result
    })
}
```

So when you want to handle an error you explicitly say so, but if you do not want to handle it or it is unexpected it will go to the catch :)

Does that make sense?

**Comment by fopsdev** [comment]
we are only speaking of sequence here. but how to handle parallel(...)?
please have a look at the tutorial bin: https://www.webpackbin.com/bins/-KdBaa45GzVJFOxU69Gp
there you can find a  usecase for errorhandling (how its done today) inside parallel in controller.js, how would we solve that?

**Comment by christianalfoni** [comment]
@fopsdev It would be the same :)

```js
parallel([

], [

])
```

When an exception occurs, and some other action in parallel resolves, it will just stop... the result of "the other actions" will still be resolved though, as you can not cancel a promise :)

### A suggestion on wording

When we discuss this I think we need to separate the words a bit. I think we should name `exceptions` the unexpected stuff that happens in your code. Basically when you `throw`, it being the runtime itself or you explicitly. This is the stuff that goes down the `exception` chain.

When we talk about `error`, that is not an `exception`. We might want to handle errors. For example bad username is something we want to handle, but it is an error. But we can treat that error as an exception by throwing to the exception chain :)

**Comment by christianalfoni** [comment]
### A suggestion on wording

When we discuss this I think we need to separate the words a bit. I think we should name `exceptions` the unexpected stuff that happens in your code. Basically when you `throw`, it being the runtime itself or you explicitly. This is the stuff that goes down the `exception` chain.

When we talk about `error`, that is not an `exception`. We might want to handle errors. For example bad username is something we want to handle, but it is an error. But we can treat that error as an exception by throwing to the exception chain :)

**Comment by gaspard** [comment]
Let's take back our use case:

```js
export const  mounted = sequence([
  setupDeviceSettings,
  setAuthHeaders,
  fetchCurrentUser, 
  set(props`currentUser`, state`currentUser`),
  set(state`currentUser.current_account_id`, state`navbar.newPost.account_id`),
  fetchPosts,
  set(props`posts`, state`posts`),
  when(state`flags.showCommentsButton`), {
    true: [
      fetchComments,
      set(props`comments`, state`comments`)
    ],
    false: [/*noop*/]
  }
], [
  // This should only happen with failures chosen by the developer, not runtime errors
  flash('Error fetching data')
])
```

I think this works. But let's not catch all exceptions, at least not at this level. The exception handling code here is for **expected exceptions**, not **whatever bad happens**.

Any generic exception (not a typeof `sequence.abort`) should bubble all the way up where it could get caught in production to log the problem or it could end up in the debugger.

**Comment by christianalfoni** [comment]
Hm, yeah... this is a difficult one. Cause:

1. We can adjust to existing solutions, like Promises and Observables... where concept of "catch" is related to throwing errors

2. We can create our own version of `execptions`

I agree with you that catching runtime exceptions not related to business logic is really annoying, but at the same time I think it is expected. Because that is how promises and observables work.

Hm hm hm... the way you would handle this in promises/obsverable land would be to throw again:

```js
export const  mounted = sequence([
  setupDeviceSettings,
  ...
], [
  flash('Error fetching data'),
  throwError
])
```
You throw it from the catcher.

Hm hm hm... thinking :)

**Comment by henri-hulski** [comment]
Yeah creating a custom exception class which extends Error and while catching check if error is instance of this class, else through it again.

**Comment by henri-hulski** [comment]
After it is also possible to create individual error classes which extends our custom class. 
This allows for more fine-grained error catching and handling.

**Comment by gaspard** [comment]
@christianalfoni I think the pseudo-code to handle this in function-tree would be something like:

```js
try {
  run(sequence)
} catch (err) {
  if (err instanceof SequenceAbort && sequence.abortPath) {
    run(sequence.abortPath)
  } else {
    throw err
  }
}
```

**Comment by henri-hulski** [comment]
@gaspard +1 maybe call it just Abort, as it could also be used for parallel.

**Comment by christianalfoni** [comment]
Hm, yeah... I would love to "fix the issue" of business logic exceptions and JS exceptions. Since function-tree is an abstraction over Promises... it is about business logic, it makes sense to do it this way indeed :)

I agree with @henri-hulski it needs to be general `Abort`.

But okay, one step further. Now we need to land on the "bubbling".

So objectively bubbling is implicit. You will have to jump between files to understand what gets caught where. Also with bubbling we would have to introduce a "keep throwing up". In case it is caught, but you want to throw the error further "up the tree".

This can become a bit problematic as Cerebral is highly composable. Meaning that some sequence of actions tries to "rethrow error up the tree" where there is no handler in one composition, but there is a handler in a different composition.

At the same time we should allow composing sequences together and have "one exception catcher to rule them all", like:

```js
sequence([
  someAction,
  someOtherSequence,
  andAnotherSequence
], [
  showError
])
```

So I suggest we implement a first version without this "explicitly throw it up the tree". When you abort it will find the nearest catch chain and if there is no chain there it will trigger the `abort` event.

This might lead to a total refactor of function-tree though. It is getting very complex now and I have no clear image of how to implement this. Will require some research and hopefully find a good way to implement :)

**Comment by bfitch** [comment]
> So I suggest we implement a first version without this "explicitly throw it up the tree".

That's exactly what I was going to recommend :)

**Comment by maxfrigge** [comment]
A bit late here :) .. but love the idea of default paths. 

What I don't get is.. why not use the initial suggestion to do

```js
sequence([
], {
  defaultPath1: [],
  error: [] // handled error returned from action
})
```

It will be more flexible and cause less breaking changes in the future using the second argument as a special task for exceptions. You can still have a "special paths" for those cases. This will also not force you to throw errors.. and rather return an error path if you prefer.

```js
sequence([
], {
  defaultPath1: [],
  error: [],
  uncaughtException: [],
  unhandledRejection: []
})
```

**Comment by christianalfoni** [comment]
@maxfrigge 

The thing is that there are no errors or exceptions in function-tree now. There are only paths. It is just a "coincidence" that some of them are called "error" :)

Secondly to have default paths means that all actions will need to be injected with default paths behind them, meaning that we will overload the debugger with a lot of information, information that mostly is never used.

Last but not least... it does not help readability, it becomes implicit :) Using default paths will remove the link between an action and a path it uses. Because now any action inside your tree could be going down any default path... you would have to look inside every action file to figure out what it uses.

`Abort` is something different cause it aborts signal execution and is considered an `exception`, not "handle this response/state as an error path" :)

**Comment by maxfrigge** [comment]
@christianalfoni for the sakes of not being the one of questioning things all the time :) I will just say that there should also be a way to end function-tree execution in "a good way". So maybe `execution.end(payload)` and `execution.fail(payload)`.

So this means that the new special path will only be triggered on explicit execution abort? Uncaught exception as well as unhandled rejections will no go down that route?

**Comment by christianalfoni** [comment]
Haha ;-)

Now I feel difficult, but is not "a good way" when the tree finishes its execution as defined? :)

Yeah, exactly, so on low level then something bad happens you would "throw new Error()", but since the JS engine does the same thing we want to separate JS engine errors and business logic errors. So `abort` is the business logic version of an error :) Any "throw", explicit or implicit, will cause an "error" event as normal.

**Comment by gaspard** [comment]
@maxfrigge I was one of the main proponents of default paths but I agree with Christian: it looks very nice and powerful in theory but it is adding quite some complexity for the "next coder" depending on how it is used so we are now focusing on **errors** or how we now call them **exceptions**.

This "smaller goal" has the advantage of not messing the readability too much.

@christianalfoni I would not bubble up. An error path belongs to the sequence where it is declared. I see "sequence" as a black box with it's execution and exception path.

For me an error is either caught right next to where it happens (in the sequence) or at the top-most level (to avoid production app crash and show some "ouups" error).

**Comment by christianalfoni** [comment]
@gaspard Great, we need to do that first anyways. If bubbling shows itself useful it can be added later :)

**Comment by maxfrigge** [comment]
@christianalfoni when using function-tree for signals in cerebral there might not be a good abort. But when using function-tree in a different environment there may be cases where you want to end early. I have such a case in my chat bot with function-tree project .. but yes this might not be the intended use .. just saying :)

@gaspard well said! so the takeaway is.. keep the pyramid of doom in favour of better visibility - is that about right?

**Comment by bfitch** [comment]
@maxfrigge no pyramid of doom! Just restricting "jumping out" of normal signal execution to explicitly thrown exceptions vs the more general purpose (and powerful?) idea of jumping to arbitrarily defined top level paths.

**Comment by maxfrigge** [comment]
@bfitch so you can still build around the happy path and instead returning error path you abort.. so less error paths in the signals?

**Comment by bfitch** [comment]
so it's really about cleaning up signals by handling errors as exceptions vs handling them...over and over again...with `{error: []}` paths. And also you only need to define a top level "catch" as the second argument to `sequence`

**Comment by christianalfoni** [comment]
@maxfrigge Yeah, lets look more into that using the chatbot as example. Would love to look more deeply into how you solve that :) Maybe demo on next hangout?

Yeah, pyramid of doom, should be explained how this new thing solves that :) Typically you would:

```js
[
  doThis, {
    success: [
      doThat, {
        success: [],
        error: []
      }
    ],
    error: []
  }
]
```

Especially on server side you typically do not want to "handle the errors". You just want to abort the whole thing when it occurs. So first iteration you might think our new approach allows us to do:


```js
sequence([
  doThis, {
    success: [
      doThat, {
        success: []
      }
    ]
  }
], [

])
```

But actually it resolves to this:

```js
sequence([
  doThis,
  doThat
], [

])
```

The reason is that you no longer have to explicitly state "the happy path", the happy path is... "just continue" :)

**Comment by maxfrigge** [comment]
@christianalfoni aaa finally got it :) that was exactly what I was thinking. Didn't realise that the abort will abort :-P

Can show the function-tree stuff I am working on in a hangout. But wouldn't you agree that abort is somewhat in the middle of good and bad.

**Comment by christianalfoni** [comment]
Hehe, yeah, abort is implicit... like, by reading the code you never know which one can cause an `abort` to happen. That said, it cleans up so well for these types of scenarios :)

Ah, man, just got very happy. The "merge execution" logic is perfect for handling ABORT :-) Cause it is already generic. It basically is a signal that has a "executedBy" prop. Meaning that an abort can fire off a new function tree execution, only it attaches the "executedBy" prop. It will automatically get inlined on the exact action that caused it... do not have to do anything :) Only thing is identifying it is an ABORT execution and style it.

Woop woop 🎉

**Comment by christianalfoni** [comment]
hah, awesome... it worked on first try... implementation driven development :-D Implement first, tes after!

**Comment by christianalfoni** [comment]
Okay, this is now implemented, also in debugger. Glad we got to this conclusion as it "felt right" implementing as well :)

I will just write some docs on it tomorrow and then create PR. This is how it looks now:

<img width="625" alt="screen shot 2017-03-21 at 22 11 05" src="https://cloud.githubusercontent.com/assets/3956929/24171080/542c8926-0e83-11e7-8157-8b7c96487b6a.png">

I tried with a red color, but it felt wrong as it is not like "something you do not want"... so I kept the same styling as when you "merge execution" client/server. Let me know if you have any ideas to indicating it differently :) But yeah.. no it is displayed on the action that actually aborts.

I also added an `AbortProvider`, meaning you can just do:

```js
function someAction ({abort}) {
  return abort()
}
```

**Comment by Guria** [comment]
How about not overloading sequence/parallel, but use chaining instead as in then/catch pattern?

```js
sequence([
  doThis,
  doThat
])
  .abort([])

parallel([
  doThis,
  doThat
])
  .abort([])
```

**Comment by gaspard** [comment]
Good idea @Guria. This also helps readability and clarifies what this "second chain" is for.

**Comment by christianalfoni** [comment]
@Guria @gaspard 

I actually tried that :)

1. I think it would look nice if it was at "signal definition", but because you can do this:

```js
sequence([
  actionA,
  parallel([
    actionB,
    actionC
  ]).abort([

  ])
])
  .abort([

  ])
```

It introduces new type of syntax "inside" of chains as well. We could support both though. Just like promises does. You can pass two callbacks to "then", one being the error handler.

**Comment by christianalfoni** [comment]
Actually, thinking about it. `.abort` hints to a "chainable API", which sequence/parallel is not. I think we give the wrong impression of how it works by introducing chainable API when it is not :) Agree it looks nice in the example, but it says more than "look nice", it says something about what kind of API it is... and it is not chainable. At least, that is a point of view to consider :-D

**Comment by Guria** [comment]
okay

**Comment by christianalfoni** [comment]
Okay, so been lots of discussions today and I have been thinking about `function-tree` low level nature. With Cerebral we have many ways to handle errors, but `function-tree` should probably not innovate too much ;-)

What I specifically mean is that `function-tree` should probably work like Promises and Observables. You attach "catch" sequences and those are run when an error is thrown. And if there is no catch sequence, it bubbles up to find one. If none is find an `error` event is emitted.

I agree with @gaspard that native JS errors are annoying to get into your business logic "catches", but I think it is even more important to have familiar behaviour. We can just say: `function-tree` works like promises and observables. It has `catch` logic on throwing that bubbles up.

So in code you would do this:

```js
sequence([
  function someFunc () {
    throw 'something'
  }
], [
  // current payload with merged in {error: 'something'}
])
```

Same with promise:

```js
sequence([
  function someFunc () {
    return new Promise((resolve, reject) => reject('something'))
  }
], [
  // current payload with merged in {error: 'something'}
])
```

That would mean there is no special "abort" in `function-tree`. The debugger would rather highlight the action with RED as normal, but instead of showing code the failed, it would show the execution of the `catch` sequence.

Been twirling around in my head and I think at this level we need to be more familiar. We can rather do stuff in Cerebral to make error handling more approachable. Has already been suggestions on allowing modules to handle errors, we could expose error operators like:

```js
sequence([

], whenError({
  'SyntaxError': [],
  'MyAppError': []
})
```

Where it would otherwise throw the error further "up" :)

**Comment by christianalfoni** [comment]
Okay, thinking more. I think it will be a mistake catching errors:

1. There was a huge problem with Promises when they arrived cause they would eat up errors. Browsers finally implemented "uncaught promise rejection" and things got better, but like in Node it is a complete mess, no strack trace on those (at least to my knowledge). Anyways, the problem is that now we eat up all errors that are not explicitly handled or rethrown

2. Errors are not serializable... we want to give a payload to the `catch sequence`. The error itself can not be the payload, cause that would give problems with existing payload we want to pass. We could hardcode it to be on the `error` property, but still... serialized errors will just be "{}"

3. Errors can not pass object of information, just a string. But we want to bail out of execution and pass information to the `catch sequence`, just as if you returned any payload from a function

4. Even errors related to React would be eaten up, because a signal synchronously triggers a flush to the UI running code there... any faulty code could end up in the `catch sequence`, being eaten up

The current approach solves all of this. We do not eat errors you want to know about, we do not have to make some opinion and spray fairy dust on making errors serializable and now we can pass any payload to the `catch sequence`.

**Comment by fopsdev** [comment]
I was also thinking that throwing an error with payload could be beneficial. 
Beeing still a bit nooby with js i'm so thankful for this issue which clarified quite a few things. No (serializable) payload on error -> no go :)

**Comment by maxfrigge** [comment]
I thought the intention was to let uncaught errors do there thing and only handle `abort(payload)` with that second chain - what happened to that? :)

**Comment by fopsdev** [comment]
@maxfrigge reading from christians post above. ```catch sequence``` === your second chain . so nothing happened to that :) 
can't wait to testdrive the PR, but got other work on the table...

**Comment by maxfrigge** [comment]
@fopsdev I meant the idea of not catching errors but only handling aborts.

**Comment by fopsdev** [comment]
oh no, anyways,  looks like i'm confusing cerebral and f-t. so i'll keep :no_mouth: now :) well, maybe a last /play yodel here...

**Comment by henri-hulski** [comment]
I think the discussion happens at the moment mainly in the discord release channel.

**Comment by gaspard** [comment]
@christianalfoni#1702 I lost the some steps. You are now working on catching errors at signal level ?

I would use this:

```js
// ==== sequence definition without abort
export const foo = [
  actionA,
  actionB,
  actionC
]

// or (equivalent)
export const foo = sequence([
  actionA,
  actionB,
  actionC
])

// ==== sequence with abort
export const foo = sequence([
  actionA,
  actionB,
  actionC
], {
  abort: [
    // abort path
  ]
})

// An action is aborted with
export function ActionA ({abort}) {
  abort({some: 'payload'})
}
```

And this for errors (and uncaught abort?) at the signal level:

```js
export const someSignal = sequence(
  [], {
    catch: [
      // error handling path*
    ]
  }
)
```

So basically, "sequence" has two parameters: a list of actions and/or sequences and an "options" object.

A catch that is not at the top level raises an error during tree compilation.

An "aborted" action without an "abort" path is handled at the top level in the "catch".

**Comment by gaspard** [comment]
And finally, the controller definition contains a "catch" option to handle uncaught errors at the top-most level.

```js
const controller = Controller({
  // Defines the top level state
  state: {},

  // Defines the top level signals
  signals: {},

  // Defines the top level modules
  modules: {}

  // Sequence to handle uncaught errors in signals
  catch: []
})
```

**Comment by fopsdev** [comment]
@gaspard really like that proposal
just to clarify:
```js
// ==== sequence definition without abort
export const sequenceWithoutAbortPath = [
  actionA,
  actionThatAborts,
  actionC
]

// ==== sequence with abort
export const sequenceWithAbortPath = sequence([
  actionA,
  actionB,
  sequenceWithoutAbortPath,
  actionC
], {
  abort: [
    // abort path
  ]
})
```
will the ActionThatAborts be catched in the outer sequence?

**Comment by christianalfoni** [comment]
https://github.com/cerebral/cerebral/pull/793

**Comment by edgesoft** [comment]
@christianalfoni  Can this be closed now?
